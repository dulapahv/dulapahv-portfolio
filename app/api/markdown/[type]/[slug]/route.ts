import { readFile } from 'fs/promises';
import { join } from 'path';
import { notFound } from 'next/navigation';

import matter from 'gray-matter';

import { BASE_URL } from '@/lib/constants';
import {
  getCollection,
  isValidContentType,
  type ContentType,
} from '@/lib/content-utils';

export async function GET(
  req: Request,
  ctx: RouteContext<'/api/markdown/[type]/[slug]'>,
) {
  const { type, slug } = await ctx.params;

  if (!isValidContentType(type)) {
    notFound();
  }

  try {
    // Read and parse the MDX file
    const filePath = join(
      '/opt/buildhome/repo/',
      'content',
      type,
      `${slug}.mdx`,
    );
    const fileContent = await readFile(filePath, 'utf-8');

    // Parse frontmatter and content
    const { data: frontmatter, content: bodyContent } = matter(fileContent);

    // Get additional metadata from content-collections
    const collection = getCollection(type as ContentType);
    const page = collection.find((p) => p._meta.path === slug);

    if (!page) {
      notFound();
    }

    // Build markdown with parsed data
    const title = `${frontmatter.title || frontmatter.position || 'Untitled'}`;
    let mdContent = `# ${title}\n\n`;

    if (frontmatter.description) {
      mdContent += `> ${frontmatter.description}\n\n`;
    }

    mdContent += `**Type:** ${type.charAt(0).toUpperCase() + type.slice(1)}\n`;

    if (frontmatter.company) {
      mdContent += `**Company:** ${frontmatter.company}\n`;
      mdContent += `**Location:** ${frontmatter.location}\n`;
    }

    const formatDate = (dateStr: string) => {
      const [day, month, year] = dateStr.split('-').map(Number);
      const date = new Date(year, month - 1, day);

      return date.toLocaleDateString('en-US', {
        month: 'long',
        day: 'numeric',
        year: 'numeric',
      });
    };

    if (frontmatter.date) {
      mdContent += `**Published:** ${formatDate(frontmatter.date)}\n`;
    } else if (frontmatter.startDate) {
      const endDate = frontmatter.endDate || 'Present';
      mdContent += `**Duration:** ${formatDate(frontmatter.startDate)} - ${
        endDate === 'Present' ? endDate : formatDate(endDate)
      }\n`;
    }

    mdContent += `**URL:** ${BASE_URL}/${type}/${slug}\n`;
    mdContent += `**Accessed:** ${new Date().toLocaleDateString('en-US', {
      month: 'long',
      day: 'numeric',
      year: 'numeric',
    })}\n`;

    let coverImageAltText = '';
    if (type === 'work') {
      coverImageAltText = `Cover image for ${frontmatter.position} position at ${frontmatter.company}`;
    } else if (type === 'project') {
      coverImageAltText = `Cover image for ${frontmatter.title} project`;
    } else {
      coverImageAltText = `Cover image for blog post: ${frontmatter.title}`;
    }

    if (frontmatter.image) {
      const imageUrl = frontmatter.image.startsWith('http')
        ? frontmatter.image
        : `${BASE_URL}${frontmatter.image}`;
      mdContent += `\n![${coverImageAltText}](${imageUrl})\n`;
    }

    mdContent += '\n---\n';
    mdContent += bodyContent;

    return new Response(mdContent, {
      headers: {
        'Content-Type': 'text/plain; charset=utf-8',
        'Cache-Control': 'public, max-age=0, must-revalidate',
        'Content-Disposition': `inline; filename="${slug}.md"`,
      },
    });
  } catch (error) {
    console.error('Error loading markdown:', error);
    notFound();
  }
}
