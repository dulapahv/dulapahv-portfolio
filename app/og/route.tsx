import { ImageResponse } from 'next/og';
import type { NextRequest } from 'next/server';

import { readFile } from 'node:fs/promises';
import { join } from 'node:path';

export const GET = async (request: NextRequest) => {
  const getParam = (key: string): string | null => {
    return request.nextUrl.searchParams.get(key) || request.nextUrl.searchParams.get(`amp;${key}`);
  };

  const title = getParam('title');
  const description = getParam('description');

  const logo = await readFile(join(process.cwd(), 'app/og/logo.png'));
  const logoSrc = `data:image/png;base64,${logo.toString('base64')}`;
  const geistBold = await readFile(join(process.cwd(), 'app/og/Geist-Bold.ttf'));
  const geistRegular = await readFile(join(process.cwd(), 'app/og/Geist-Regular.ttf'));

  return new ImageResponse(
    (
      <div
        tw="flex flex-col justify-between items-start w-full h-full bg-black p-12"
        style={{
          backgroundSize: '80px 80px',
          backgroundImage:
            'linear-gradient(to right, #1C1C1C 1px, transparent 1px), linear-gradient(to bottom, #1C1C1C 1px, transparent 1px)'
        }}
      >
        <img src={logoSrc} alt="avatar" width={72} height={72} tw="overflow-hidden" />
        <div tw="flex flex-col">
          <h1 tw="max-w-[48rem] text-[64px] font-bold leading-[69px] tracking-tighter m-0 text-[#F1F1F1]">
            {title}
          </h1>
          {description && (
            <p tw="max-w-[64rem] text-[24px] font-normal leading-[32px] tracking-tight text-[#A5A5A5] mt-4 mb-0">
              {description}
            </p>
          )}
        </div>
      </div>
    ),
    {
      width: 1200,
      height: 630,
      fonts: [
        {
          name: 'Geist',
          data: geistBold,
          style: 'normal',
          weight: 700
        },
        {
          name: 'Geist',
          data: geistRegular,
          style: 'normal',
          weight: 400
        }
      ]
    }
  );
};
