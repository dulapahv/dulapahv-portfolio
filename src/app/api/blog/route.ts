import { NextResponse } from "next/server";

import prisma from "@/app/db";

export async function GET(request: Request) {
  const response = await prisma.blog.findMany();

  return NextResponse.json(
    {
      response,
    },
    {
      status: 200,
    },
  );
}
