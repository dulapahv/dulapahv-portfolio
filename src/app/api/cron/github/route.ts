import { Octokit } from "@octokit/rest";

import { SHORT_NAME } from "@/lib/constants";
import { updateEdgeConfig } from "@/lib/vercel";
import { parseError } from "@/utils/parse-error";

const octokit = new Octokit();

export const GET = async (): Promise<Response> => {
  try {
    const profile = await octokit.rest.users.getByUsername({
      username: SHORT_NAME,
    });

    await updateEdgeConfig("github", {
      followers: profile.data.followers,
    });

    return new Response(undefined, { status: 204 });
  } catch (error) {
    const message = parseError(error);

    return new Response(message, { status: 500 });
  }
};
