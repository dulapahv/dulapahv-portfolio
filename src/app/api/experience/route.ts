import prisma from "@/lib/prisma";

export async function GET(request: Request) {
  const data = await prisma.experience.findMany();

  return Response.json({ data });
}
