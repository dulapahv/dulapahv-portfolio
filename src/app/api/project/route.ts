import { NextResponse } from "next/server";

import prisma from "@/app/db";

export async function GET(request: Request) {
  const response = await prisma.project.findMany();

  return NextResponse.json(
    {
      response,
    },
    {
      status: 200,
    },
  );
}
