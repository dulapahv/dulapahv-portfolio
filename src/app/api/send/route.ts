import { NextRequest } from "next/server";
import { ErrorResponse, Resend } from "resend";

import { NAME } from "@/lib/constants";
import { EmailTemplateProps } from "@/types/types";
import { EmailTemplate } from "@/ui/email-template";
import { parseError } from "@/utils/parse-error";

export const runtime = "edge";

const resend = new Resend(process.env.RESEND_API_KEY);

export async function POST(req: NextRequest) {
  const fullName = req.nextUrl.searchParams.get("name");
  const email = req.nextUrl.searchParams.get("email");
  const type = req.nextUrl.searchParams.get("type");
  const message = req.nextUrl.searchParams.get("message");

  if (!fullName || !email || !type || !message) {
    const error = {
      message: "Please fill in all required fields",
      name: "missing_required_field",
    } as ErrorResponse;
    return Response.json({ error }, { status: 422 });
  }

  try {
    const { data, error } = await resend.emails.send({
      from: `${NAME} <${process.env.EMAIL}>`,
      to: [`${process.env.EMAIL}`],
      subject: `âœ¨ ${type} - ${fullName}`,
      react: EmailTemplate({
        fullName,
        email,
        type,
        message,
      } as EmailTemplateProps),
    });
    if (error) {
      return Response.json({ error }, { status: 500 });
    }
    return Response.json(data);
  } catch (error) {
    const message = parseError(error);

    console.error(message);

    return Response.json(message, { status: 500 });
  }
}
