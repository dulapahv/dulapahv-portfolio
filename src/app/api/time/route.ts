import { parseError } from "@/utils/parse-error";

export const GET = async (): Promise<Response> => {
  try {
    const res = await fetch(
      "http://worldtimeapi.org/api/timezone/Europe/London",
      {
        cache: "no-store",
      },
    );
    const data = await res.json();
    return Response.json(data);
  } catch (error) {
    const message = parseError(error);

    console.error(message);

    return new Response(message, { status: 500 });
  }
};
