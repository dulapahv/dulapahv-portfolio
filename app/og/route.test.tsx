import { NextRequest } from 'next/server';

import { beforeEach, describe, expect, it, vi } from 'vitest';

import { GET } from './route';

vi.mock('next/og', () => ({
  ImageResponse: class MockImageResponse {
    constructor(
      /* eslint-disable  @typescript-eslint/no-explicit-any */
      public element: any,
      /* eslint-disable  @typescript-eslint/no-explicit-any */
      public options: any
    ) {}
  }
}));

vi.mock('node:fs/promises', () => {
  const readFile = vi.fn();
  const writeFile = vi.fn();
  return {
    readFile,
    writeFile,
    default: { readFile, writeFile },
    __mocks__: { readFile, writeFile }
  };
});

vi.mock('node:path', () => {
  const join = vi.fn((...args: string[]) => args.join('/'));
  return {
    join,
    default: { join },
    __mocks__: { join }
  };
});

/* eslint-disable  @typescript-eslint/no-explicit-any */
const { __mocks__: fsMocks } = vi.mocked(await import('node:fs/promises')) as any;
/* eslint-disable  @typescript-eslint/no-explicit-any */
const { __mocks__: pathMocks } = vi.mocked(await import('node:path')) as any;

const mockReadFile = fsMocks.readFile;
const mockJoin = pathMocks.join;

describe('OG Image Route', () => {
  beforeEach(() => {
    vi.clearAllMocks();
    // Setup default mock implementations
    mockReadFile.mockResolvedValue(Buffer.from('mock-file-content'));
    mockJoin.mockImplementation((...args: string[]) => args.join('/'));
  });

  it('should generate OG image with title and description', async () => {
    const url = new URL('https://example.com/og?title=Test%20Title&description=Test%20Description');
    const request = new NextRequest(url);

    const response = await GET(request);

    expect(response).toBeDefined();
    expect(response).toBeInstanceOf(Object);
  });

  it('should generate OG image with only title (no description)', async () => {
    const url = new URL('https://example.com/og?title=Test%20Title');
    const request = new NextRequest(url);

    const response = await GET(request);

    expect(response).toBeDefined();
  });

  it('should handle amp-encoded parameters', async () => {
    const url = new URL(
      'https://example.com/og?amp;title=Test%20Title&amp;description=Test%20Description'
    );
    const request = new NextRequest(url);

    const response = await GET(request);

    expect(response).toBeDefined();
  });

  it('should load font files', async () => {
    const url = new URL('https://example.com/og?title=Test');
    const request = new NextRequest(url);

    await GET(request);

    // Check that readFile was called for logo and fonts
    expect(mockReadFile).toHaveBeenCalledWith(expect.stringContaining('logo.png'));
    expect(mockReadFile).toHaveBeenCalledWith(expect.stringContaining('Geist-Bold.ttf'));
    expect(mockReadFile).toHaveBeenCalledWith(expect.stringContaining('Geist-Regular.ttf'));
  });

  it('should return ImageResponse with correct dimensions', async () => {
    const url = new URL('https://example.com/og?title=Test');
    const request = new NextRequest(url);

    const response = await GET(request);

    // @ts-expect-error - accessing options for testing
    expect(response.options.width).toBe(1200);
    // @ts-expect-error - accessing options for testing
    expect(response.options.height).toBe(630);
  });

  it('should configure fonts correctly', async () => {
    const url = new URL('https://example.com/og?title=Test');
    const request = new NextRequest(url);

    const response = await GET(request);

    // @ts-expect-error - accessing options for testing
    const fonts = response.options.fonts;
    expect(fonts).toHaveLength(2);
    expect(fonts[0]).toMatchObject({
      name: 'Geist',
      style: 'normal',
      weight: 700
    });
    expect(fonts[1]).toMatchObject({
      name: 'Geist',
      style: 'normal',
      weight: 400
    });
  });

  it('should handle empty parameters', async () => {
    const url = new URL('https://example.com/og');
    const request = new NextRequest(url);

    const response = await GET(request);

    expect(response).toBeDefined();
  });

  it('should handle special characters in title', async () => {
    const url = new URL('https://example.com/og?title=Test%20%26%20Special%20%3CCharacters%3E');
    const request = new NextRequest(url);

    const response = await GET(request);

    expect(response).toBeDefined();
  });

  it('should handle very long title', async () => {
    const longTitle = 'A'.repeat(200);
    const url = new URL(`https://example.com/og?title=${encodeURIComponent(longTitle)}`);
    const request = new NextRequest(url);

    const response = await GET(request);

    expect(response).toBeDefined();
  });

  it('should handle very long description', async () => {
    const longDescription = 'B'.repeat(500);
    const url = new URL(
      `https://example.com/og?title=Test&description=${encodeURIComponent(longDescription)}`
    );
    const request = new NextRequest(url);

    const response = await GET(request);

    expect(response).toBeDefined();
  });

  it('should match OG image snapshot', async () => {
    const url = new URL(
      'https://example.com/og?title=Snapshot%20Test&description=Snapshot%20Description'
    );
    const request = new NextRequest(url);

    const response = await GET(request);

    expect(response).toMatchSnapshot();
  });
});
