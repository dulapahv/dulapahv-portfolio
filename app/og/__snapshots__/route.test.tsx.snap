// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OG Image Route > should match OG image snapshot 1`] = `
MockImageResponse {
  "element": <div
    style={
      {
        "backgroundImage": "linear-gradient(to right, #1C1C1C 1px, transparent 1px), linear-gradient(to bottom, #1C1C1C 1px, transparent 1px)",
        "backgroundSize": "80px 80px",
      }
    }
    tw="flex flex-col justify-between items-start w-full h-full bg-black p-12"
  >
    <img
      alt="avatar"
      height={72}
      src="data:image/png;base64,bW9jay1maWxlLWNvbnRlbnQ="
      tw="overflow-hidden"
      width={72}
    />
    <div
      tw="flex flex-col"
    >
      <h1
        tw="max-w-[48rem] text-[64px] font-bold leading-[69px] tracking-tighter m-0 text-[#F1F1F1]"
      >
        Snapshot Test
      </h1>
      <p
        tw="max-w-[64rem] text-[24px] font-normal leading-[32px] tracking-tight text-[#A5A5A5] mt-4 mb-0"
      >
        Snapshot Description
      </p>
    </div>
  </div>,
  "options": {
    "fonts": [
      {
        "data": {
          "data": [
            109,
            111,
            99,
            107,
            45,
            102,
            105,
            108,
            101,
            45,
            99,
            111,
            110,
            116,
            101,
            110,
            116,
          ],
          "type": "Buffer",
        },
        "name": "Geist",
        "style": "normal",
        "weight": 700,
      },
      {
        "data": {
          "data": [
            109,
            111,
            99,
            107,
            45,
            102,
            105,
            108,
            101,
            45,
            99,
            111,
            110,
            116,
            101,
            110,
            116,
          ],
          "type": "Buffer",
        },
        "name": "Geist",
        "style": "normal",
        "weight": 400,
      },
    ],
    "height": 630,
    "width": 1200,
  },
}
`;
